topdown/json.go

```diff
@@ -20,65 +124,75 @@ func builtinJSONFilter(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Te
 	}
 
 	// Build a list of filter strings
-	var filters [][]ast.Value
+	filters, err := getJSONPaths(operands[1].Value)
+	if err != nil {
+		return err
+	}
+
+	// Actually do the filtering
+	filterObj := pathsToObject(filters)
+	r, err := obj.Filter(filterObj)
+	if err != nil {
+		return err
+	}
 
-	switch v := operands[1].Value.(type) {
+	return iter(ast.NewTerm(r))
+}
+
+func getJSONPaths(operand ast.Value) ([]ast.Ref, error) {
+	var paths []ast.Ref
+
+	switch v := operand.(type) {
 	case ast.Array:
 		for _, f := range v {
 			filter, err := parsePath(f)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			filters = append(filters, filter)
+			paths = append(paths, filter)
 		}
 	case ast.Set:
 		err := v.Iter(func(f *ast.Term) error {
 			filter, err := parsePath(f)
 			if err != nil {
 				return err
 			}
-			filters = append(filters, filter)
+			paths = append(paths, filter)
 			return nil
 		})
 		if err != nil {
-			return err
+			return nil, err
 		}
 	default:
-		return builtins.NewOperandTypeErr(2, v, "set", "array")
+		return nil, builtins.NewOperandTypeErr(2, v, "set", "array")
 	}
 
-	// Actually do the filtering
-	filterObj := pathsToObject(filters)
-	r, err := obj.Filter(filterObj)
-	if err != nil {
-		return err
-	}
-
-	return iter(ast.NewTerm(r))
+	return paths, nil
 }
 
-func parsePath(path *ast.Term) ([]ast.Value, error) {
+func parsePath(path *ast.Term) (ast.Ref, error) {
 	// paths can either be a `/` separated json path or
 	// an array or set of values
-	var pathSegments []ast.Value
+	var pathSegments ast.Ref
 	switch p := path.Value.(type) {
 	case ast.String:
 		parts := strings.Split(strings.Trim(string(p), "/"), "/")
 		for _, part := range parts {
-			pathSegments = append(pathSegments, ast.String(part))
+			part = strings.ReplaceAll(strings.ReplaceAll(part, "~1", "/"), "~0", "~")
+			pathSegments = append(pathSegments, ast.StringTerm(part))
```

Comment:
Similar to above, can we just use the array literal option for this case?

Reply:
Not sure what you mean. This is taking a bunch of strings and making a ref out of them. The reason it changed from an array of values is that we kept having to make terms to reference into objects/sets and the `nil` checking was kind of a hassle. Having them just be terms from the get-go simplified a bunch of stuff.

Reply:
I thought we had a similar problem with another built-in function and the solution was to let callers specify a string (e.g., /foo/bar/baz) or an array (["foo", "bar", "baz"]). The array option doesn't have the same escaping problem, etc. Not a big deal but something I figured could simplify implementation a bit.

Reply:
Ohhh, you're talking about the escaped characters. This was just to make it support json pointer paths more fully (https://tools.ietf.org/html/rfc6901#section-3) I don't really have strong feelings about it. Just makes it more json pointer-y for folks that are used to using them.